### PR TITLE
#330 add checks for metrics

### LINKED ISSUE #330: Add checks for metrics
Would be good to add additional checks for the metrics.
For example:

- check that `y_true` and `y_pred` are the same length
- check that there are no Nans

Simple check to assure that the user has good feedback upon error in metric.

### PR DESCRIPTION
**Pull request by JumpingDino**
# Description
Checks are good for metrics calculation. It's a good idea to assure the vectors (y_pred, y_probs) are the same size and don't have NaNs or infs. For this, two functions were created:
- check_arrays_length
- check_array_nan

Fixes #330 

## Type of change
Create some checks on `utils.py` and these functions are used in the metrics in metrics.py

# How Has This Been Tested?
Creation of tests in `test_utils.py`
- test_length
- test_inf_values
- test_nan_values

# Checklist

- [X] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [X] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [X] Linting passes successfully : `make lint`
- [X] Typing passes successfully : `make type-check`
- [X] Unit tests pass successfully : `make tests`
- [X] Coverage is 100% : `make coverage`
- [X] Documentation builds successfully : `make doc`

### CODE REVIEW

**Comment by JumpingDino** [comment]
When I run the coverage I have the following results:
image

However, when I look at the data in `test_calibration.py`, we have some NaNs on our data
image

1. In what cases I expect to have data with NaNs? 
2. Should I refactor these np.nan? 
3. What is your recommendation?

**Comment by LacombeLouis** [comment]
Maybe we need to check that there are no Nans per row? I think that would be a good start! 😄 
We should always expect to have Nans in the calibration (Top-Label), since we only re-calibrate the top-label.

**Comment by JumpingDino** [comment]
Awesome @LacombeLouis !! Thanks for the idea. 
I implemented a function that see if the array has only NaNs.
I broke the function `check_array_nan` in `check_array_nan` and `check_array_inf`.
Thanks for the support and feel free to give feedback on code :)

File: mapie/metrics.py [hunk 1]
```diff
@@ -9,7 +9,11 @@
 from .utils import (calc_bins,
                     check_array_shape_classification,
                     check_array_shape_regression,
+                    check_array_nan,
+                    check_array_inf,
+                    check_arrays_length,
                     check_binary_zero_one,
+                    check_lower_upper_bounds,
```
Comment (vincentblot28) [inline mapie/metrics.py:16]: Please sort the imports

File: mapie/utils.py [hunk 2]
```diff
@@ -1208,3 +1208,65 @@ def check_nb_sets_sizes(sizes: NDArray, num_bins: int) -> None:
...
+
+    Raises
+    ------
+    ValueError
+        If all elements of the array are NaNs
+    """
+
+    if np.isnan(array).all() and len(np.unique(array)) > 0:
+        raise ValueError(
+            "Array contains NaN values only."
```
Comment (LacombeLouis) [inline mapie/utils.py:1229]: [suggestion]
This should say "Array contains only Nan val*"

File: mapie/utils.py [hunk 3]
```diff
@@ -1208,3 +1208,65 @@ def check_nb_sets_sizes(sizes: NDArray, num_bins: int) -> None:
...
+        If any elements of the array is +inf or -inf.
+    """
+
+    if np.isinf(array).any():
+        raise ValueError(
+            "Array contains infinite values."
+        )
+
+
+def check_arrays_length(*arrays: NDArray) -> None:
```
Comment (LacombeLouis) [inline mapie/utils.py:1254]: [question]
Why do we not the arrays to be `ArrayLike`?
If we do, then we could have an issue with `array.shape[0]`

**Review by LacombeLouis — CHANGES_REQUESTED** [review]
Thank you for the PR! It's great!

Thinking out loud here: do you think it would be smart to add a check for the metrics where we expect a score between 0 and 1. @JumpingDino @vincentblot28 @thibaultcordier

File: mapie/utils.py [hunk 4]
```diff
@@ -1208,3 +1208,65 @@ def check_nb_sets_sizes(sizes: NDArray, num_bins: int) -> None:
...
+    Raises
+    ------
+    ValueError
+        If the length of the arrays are different
+    """
+    res = [array.shape[0] for array in arrays]
+    if len(set(res)) > 1:
+        raise ValueError(
+                "There are arrays with different length"
+            )
```
Comment (thibaultcordier) [inline mapie/utils.py:1272]:
```suggestion
    res = [len(array) for array in arrays]
    if len(np.unique(res)) > 1:
        raise ValueError(
            "There are arrays with different length"
        )
```

File: mapie/metrics.py [hunk 5]
```diff
@@ -505,6 +580,7 @@ def regression_ssc_score(
     >>> print(regression_ssc_score(y_true, y_intervals, num_bins=2))
     [1.  0.5]
     """
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:583]:
```suggestion
```

File: mapie/metrics.py [hunk 6]
```diff
@@ -554,9 +630,16 @@ def classification_ssc(
     >>> print(classification_ssc(y_true, y_pred_set, num_bins=2))
     [[1.         0.66666667]]
     """
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:633]:
```suggestion
```

File: mapie/metrics.py [hunk 7]
```diff
@@ -628,6 +711,11 @@ def classification_ssc_score(
     >>> print(classification_ssc_score(y_true, y_pred_set, num_bins=2))
     [0.66666667]
     """
+    check_arrays_length(y_true, y_pred_set)
+    check_array_nan(y_true)
+    check_array_inf(y_true)
+    check_array_nan(y_pred_set)
+    check_array_inf(y_pred_set)
```
Comment (thibaultcordier) [inline mapie/metrics.py:718]:
```suggestion
    check_array_inf(y_pred_set)

```

File: mapie/metrics.py [hunk 8]
```diff
@@ -444,12 +512,19 @@ def regression_ssc(
     >>> print(regression_ssc(y_true, y_intervals, num_bins=2))
     [[1. 1.]]
     """
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:515]:
```suggestion
```

File: mapie/metrics.py [hunk 9]
```diff
@@ -221,6 +254,13 @@ def expected_calibration_error(
     y_true_ = check_binary_zero_one(y_true)
     y_scores = cast(NDArray, y_scores)
 
+    check_arrays_length(y_true_, y_scores)
+    check_array_nan(y_true_)
+    check_array_inf(y_true_)
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:260]:
```suggestion
```

File: mapie/metrics.py [hunk 10]
```diff
@@ -174,6 +204,9 @@ def classification_mean_width_score(y_pred_set: ArrayLike) -> float:
     >>> print(classification_mean_width_score(y_pred_set))
     2.0
     """
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:207]:
```suggestion
```

File: mapie/metrics.py [hunk 11]
```diff
@@ -141,6 +163,14 @@ def regression_mean_width_score(
     """
     y_pred_low = cast(NDArray, column_or_1d(y_pred_low))
     y_pred_up = cast(NDArray, column_or_1d(y_pred_up))
+
+    check_arrays_length(y_pred_low, y_pred_up)
+    check_array_nan(y_pred_low)
+    check_array_inf(y_pred_low)
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:170]:
```suggestion
```

File: mapie/metrics.py [hunk 12]
```diff
@@ -105,6 +119,14 @@ def classification_coverage_score(
             y_pred_set, force_all_finite=True, dtype=["bool"]
         )
     )
+
+    check_arrays_length(y_true, y_pred_set)
+    check_array_nan(y_true)
+    check_array_inf(y_true)
+
```
Comment (thibaultcordier) [inline mapie/metrics.py:126]:
```suggestion
```

File: mapie/utils.py (hunk 3 above)
Comment (thibaultcordier) [reply mapie/utils.py:1254]:
```suggestion
def check_arrays_length(*arrays: ArrayLike) -> None:
```

File: mapie/utils.py [hunk 13]
```diff
@@ -1208,3 +1208,65 @@ def check_nb_sets_sizes(sizes: NDArray, num_bins: int) -> None:
...
+        )
+
+
+def check_arrays_length(*arrays: NDArray) -> None:
+    """
+    Checks if the length of all arrays given in this function are the same
+
+    Parameters
+    ----------
+    *arrays: NDArray or ArrayLike
```
Comment (thibaultcordier) [inline mapie/utils.py:1260]:
```suggestion
    *arrays: ArrayLike
```

**Review by thibaultcordier — CHANGES_REQUESTED** [review]
Excellent PR, thank you for your contribution! A few suggestions to refine your code:

- Replace all `NDArray` with `ArrayLike`.
- Delete some line breaks.

You can use the functionality directly on GitHub to commit/push directly if you wish.

File: mapie/utils.py (hunk 3 above)
Comment (JumpingDino) [reply mapie/utils.py:1254]: When I change for the ArrayLike type I receive the following errors when I run `make type-check`:
```
mapie/utils.py:1268: error: Item "Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "_SupportsArray[dtype[Any]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[_SupportsArray[dtype[Any]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[Sequence[_SupportsArray[dtype[Any]]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "bool" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "int" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "float" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "complex" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "str" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "bytes" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[Union[bool, int, float, complex, str, bytes]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
mapie/utils.py:1268: error: Item "Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]" of "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]" has no attribute "shape"  [union-attr]
```

Comment (JumpingDino) [reply mapie/utils.py:1254]: Changing to @thibaultcordier proposal

```python3    
res = [len(array) for array in arrays]
    if len(np.unique(res)) > 1:
        raise ValueError(
            "There are arrays with different length"
        )
```

I had the following error: 
```
mapie/utils.py:1268: error: Argument 1 to "len" has incompatible type "Union[Sequence[Sequence[Sequence[Sequence[Sequence[Any]]]]], Union[Union[_SupportsArray[dtype[Any]], Sequence[_SupportsArray[dtype[Any]]], Sequence[Sequence[_SupportsArray[dtype[Any]]]], Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]], Sequence[Sequence[Sequence[Sequence[_SupportsArray[dtype[Any]]]]]]], Union[bool, int, float, complex, str, bytes, Sequence[Union[bool, int, float, complex, str, bytes]], Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]], Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]], Sequence[Sequence[Sequence[Sequence[Union[bool, int, float, complex, str, bytes]]]]]]]]"; expected "Sized"  [arg-type]
```

**Comment by JumpingDino** [comment]
Hi people, actually with the actual code we coverage all the steps, however when I change the `check_arrays_length` function to use ArrayLike type the check-type throws the errors said above. 

1. Could someone explain the major differences of ArrayLike and NDArray as an argument please? Maybe you can give some reference and/or ideas of benefits of using the ArrayLike type.
2. @LacombeLouis I think it makes sense to check the boundaries of the elements of the array for `classification` tasks! 
3. Do you think it makes sense to implement these sanity checks (check_nan, inf, length) in a single function? By this way I expect to have less verbosity over our metrics.py functions, what do you think?

File: mapie/utils.py [hunk 14]
```diff
@@ -1208,3 +1208,66 @@ def check_nb_sets_sizes(sizes: NDArray, num_bins: int) -> None:
...
+    Raises
+    ------
+    ValueError
+        If all elements of the array are NaNs
+    """
+
+    if np.isnan(array).all() and len(np.unique(array)) > 0:
+        raise ValueError(
+            "Array contains only NaN values."
+        )
```
Comment (thibaultcordier) [inline mapie/utils.py:1230]:
```suggestion
def check_array_nan(array: NDArray) -> None:
    """
    Checks if the array have only NaN values. If it has we throw an error.

    Parameters
    ----------
    array: NDArray
        an array with non-numerical or non-categorical values

    Raises
    ------
    ValueError
        If all elements of the array are NaNs
    """
    if np.isnan(array).all() and len(np.unique(array)) > 0:
        raise ValueError(
            "Array contains only NaN values."
        )
```

File: mapie/utils.py [hunk 15]
```diff
@@ -1208,3 +1208,66 @@ def check_nb_sets_sizes(sizes: NDArray, num_bins: int) -> None:
...
+    Raises
+    ------
+    ValueError
+        If any elements of the array is +inf or -inf.
+    """
+
+    if np.isinf(array).any():
+        raise ValueError(
+            "Array contains infinite values."
+        )
```
Comment (thibaultcordier) [inline mapie/utils.py:1251]:
```suggestion
def check_array_inf(array: NDArray) -> None:
    """
    Checks if the array have inf.
    If a value is infinite, we throw an error.

    Parameters
    ----------
    array: NDArray
        an array with non-numerical or non-categorical values

    Raises
    ------
    ValueError
        If any elements of the array is +inf or -inf.
    """
    if np.isinf(array).any():
        raise ValueError(
            "Array contains infinite values."
        )
```

File: mapie/utils.py [hunk 16]
```diff
@@ -1208,3 +1208,66 @@ def check_nb_sets_sizes(sizes: NDArray, num_bins: int) -> None:
...
+    ------
+    ValueError
+        If the length of the arrays are different
+    """
+
+    res = [len(array) for array in arrays]
+    if np.unique(res).size > 1:
+        raise ValueError(
+            "There are arrays with different length"
+        )
```
Comment (thibaultcordier) [inline mapie/utils.py:1273]:
```suggestion
    """
    Checks if the length of all arrays given in this function are the same

    Parameters
    ----------
    *arrays: NDArray
        Arrays expected to have the same length

    Raises
    ------
    ValueError
        If the length of the arrays are different
    """
    res = [len(array) for array in arrays]
    if len(np.unique(res)) > 1:
        raise ValueError(
            "There are arrays with different length"
        )
```

File: mapie/utils.py (hunk 3 above)
Comment (thibaultcordier) [reply mapie/utils.py:1254]: Indeed, it is preferable to use NDArray instead of ArrayLike.

**Review by thibaultcordier — COMMENTED** [review]
Hi @JumpingDino,
> 1. Could someone explain the major differences of ArrayLike and NDArray as an argument please? Maybe you can give some reference and/or ideas of benefits of using the ArrayLike type.

`ArrayLike`: https://numpy.org/devdocs/reference/typing.html#numpy.typing.ArrayLike
- A Union representing objects that can be coerced into an ndarray.

`NDArray`: https://numpy.org/devdocs/reference/typing.html#numpy.typing.NDArray
- Can be used during runtime for typing arrays with a given dtype and unspecified shape.

`ArrayLike` is more generic than `NDArray` except that `ArrayLike` must be cast in `NDArray` to be used. In the end, we use the parameters as an Numpy array. Sometimes, you may have to manipulate pandas type arrays, which is why we want to take into account all types of array.

When I see the error you encountered, I think we can make the assumption that the parameters are already of type `NDArray` (if this is not the case, we can simply cast it with numpy and use the `check_array` function test to check).

I've suggested a modification. You can add it directly in GitHub (you have a function that lets you commit suggestions directly in the browser).

> 3. Do you think it makes sense to implement these sanity checks (check_nan, inf, length) in a single function? By this way I expect to have less verbosity over our metrics.py functions, what do you think?

Indeed, it could make sense to implement them in a single function. But as it stands, your contribution fulfils the desired objective. You are free to make this change if you wish.

**Review by LacombeLouis — APPROVED** [review]

**Comment by JumpingDino** [comment]
I think we're good to go right?

Screenshot 2023-10-06 at 10 39 30

**Comment by thibaultcordier** [comment]
> I though about the idea to make the checks from 0 to 1 but this would require me to understand which are classification metrics and I'm kinda worried about some scenarios this may not happen. Maybe we could just finish this issue and we revisit with more advanced checks later.
>
> I think we're good to go right? What do you think?

I agree with you, we could finish this PR and open a new issue for more advanced checks.
 
> I'm available for any refinements! and thanks for the knowledge :)

Yes, just one more turn: can you add a new line in HISTORY.rst specifying that you are adding new checks, and a new line in CONTRIBUTING.rst adding your name (if you wish)? I'll be approving your PR next :)

**Review by thibaultcordier — APPROVED** [review]

**Pull request merged**